Fails to build with OpenSSL_1_1_1

The `README.md` does say that it doesn't work - but since 1.1.1 is the current standard and downgrading to 1.1.0 is both impractical and non-attractive - perhaps you can update the code to work with 1.1.1?

`master` branch fails compilation:
```
[ 37%] Building C object CMakeFiles/gost_engine.dir/gost_pmeth.c.o
/Users/ur20980/src/engine/gost_pmeth.c:62:38: error: passing 'const EVP_PKEY_CTX *' (aka 'const struct evp_pkey_ctx_st *')
      to parameter of type 'EVP_PKEY_CTX *' (aka 'struct evp_pkey_ctx_st *') discards qualifiers
      [-Werror,-Wincompatible-pointer-types-discards-qualifiers]
    src_data = EVP_PKEY_CTX_get_data((const EVP_PKEY_CTX *)src);
                                     ^~~~~~~~~~~~~~~~~~~~~~~~~
/opt/local/include/openssl/evp.h:1371:43: note: passing argument to parameter 'ctx' here
void *EVP_PKEY_CTX_get_data(EVP_PKEY_CTX *ctx);
                                          ^
/Users/ur20980/src/engine/gost_pmeth.c:573:38: error: passing 'const EVP_PKEY_CTX *'
      (aka 'const struct evp_pkey_ctx_st *') to parameter of type 'EVP_PKEY_CTX *' (aka 'struct evp_pkey_ctx_st *')
      discards qualifiers [-Werror,-Wincompatible-pointer-types-discards-qualifiers]
    src_data = EVP_PKEY_CTX_get_data(src);
                                     ^~~
/opt/local/include/openssl/evp.h:1371:43: note: passing argument to parameter 'ctx' here
void *EVP_PKEY_CTX_get_data(EVP_PKEY_CTX *ctx);
                                          ^
/Users/ur20980/src/engine/gost_pmeth.c:1052:40: error: incompatible pointer types passing 'int
      (EVP_PKEY_CTX *, const EVP_PKEY_CTX *)' (aka 'int (struct evp_pkey_ctx_st *, const struct evp_pkey_ctx_st *)') to
      parameter of type 'int (*)(EVP_PKEY_CTX *, EVP_PKEY_CTX *)' (aka 'int (*)(struct evp_pkey_ctx_st *, struct
      evp_pkey_ctx_st *)') [-Werror,-Wincompatible-pointer-types]
        EVP_PKEY_meth_set_copy(*pmeth, pkey_gost_mac_copy);
                                       ^~~~~~~~~~~~~~~~~~
/opt/local/include/openssl/evp.h:1423:35: note: passing argument to parameter 'copy' here
                            int (*copy) (EVP_PKEY_CTX *dst,
                                  ^
/Users/ur20980/src/engine/gost_pmeth.c:1062:40: error: incompatible pointer types passing 'int
      (EVP_PKEY_CTX *, const EVP_PKEY_CTX *)' (aka 'int (struct evp_pkey_ctx_st *, const struct evp_pkey_ctx_st *)') to
      parameter of type 'int (*)(EVP_PKEY_CTX *, EVP_PKEY_CTX *)' (aka 'int (*)(struct evp_pkey_ctx_st *, struct
      evp_pkey_ctx_st *)') [-Werror,-Wincompatible-pointer-types]
        EVP_PKEY_meth_set_copy(*pmeth, pkey_gost_mac_copy);
                                       ^~~~~~~~~~~~~~~~~~
/opt/local/include/openssl/evp.h:1423:35: note: passing argument to parameter 'copy' here
                            int (*copy) (EVP_PKEY_CTX *dst,
                                  ^
/Users/ur20980/src/engine/gost_pmeth.c:1072:40: error: incompatible pointer types passing 'int
      (EVP_PKEY_CTX *, const EVP_PKEY_CTX *)' (aka 'int (struct evp_pkey_ctx_st *, const struct evp_pkey_ctx_st *)') to
      parameter of type 'int (*)(EVP_PKEY_CTX *, EVP_PKEY_CTX *)' (aka 'int (*)(struct evp_pkey_ctx_st *, struct
      evp_pkey_ctx_st *)') [-Werror,-Wincompatible-pointer-types]
        EVP_PKEY_meth_set_copy(*pmeth, pkey_gost_mac_copy);
                                       ^~~~~~~~~~~~~~~~~~
/opt/local/include/openssl/evp.h:1423:35: note: passing argument to parameter 'copy' here
                            int (*copy) (EVP_PKEY_CTX *dst,
                                  ^
/Users/ur20980/src/engine/gost_pmeth.c:1083:40: error: incompatible pointer types passing 'int
      (EVP_PKEY_CTX *, const EVP_PKEY_CTX *)' (aka 'int (struct evp_pkey_ctx_st *, const struct evp_pkey_ctx_st *)') to
      parameter of type 'int (*)(EVP_PKEY_CTX *, EVP_PKEY_CTX *)' (aka 'int (*)(struct evp_pkey_ctx_st *, struct
      evp_pkey_ctx_st *)') [-Werror,-Wincompatible-pointer-types]
        EVP_PKEY_meth_set_copy(*pmeth, pkey_gost_mac_copy);
                                       ^~~~~~~~~~~~~~~~~~
/opt/local/include/openssl/evp.h:1423:35: note: passing argument to parameter 'copy' here
                            int (*copy) (EVP_PKEY_CTX *dst,
                                  ^
/Users/ur20980/src/engine/gost_pmeth.c:1091:36: error: incompatible pointer types passing 'int
      (EVP_PKEY_CTX *, const EVP_PKEY_CTX *)' (aka 'int (struct evp_pkey_ctx_st *, const struct evp_pkey_ctx_st *)') to
      parameter of type 'int (*)(EVP_PKEY_CTX *, EVP_PKEY_CTX *)' (aka 'int (*)(struct evp_pkey_ctx_st *, struct
      evp_pkey_ctx_st *)') [-Werror,-Wincompatible-pointer-types]
    EVP_PKEY_meth_set_copy(*pmeth, pkey_gost_copy);
                                   ^~~~~~~~~~~~~~
/opt/local/include/openssl/evp.h:1423:35: note: passing argument to parameter 'copy' here
                            int (*copy) (EVP_PKEY_CTX *dst,
                                  ^
7 errors generated.
make[2]: *** [CMakeFiles/gost_engine.dir/gost_pmeth.c.o] Error 1
```
 But if `-Wno-incompatible-pointer-types` is added, it compiles OK, but fails one test:
```
$ make test
Running tests...
Test project /Users/ur20980/src/engine/build
      Start  1: digest
 1/10 Test  #1: digest ...........................   Passed    0.01 sec
      Start  2: curves
 2/10 Test  #2: curves ...........................   Passed    0.03 sec
      Start  3: parameters
 3/10 Test  #3: parameters .......................   Passed    0.03 sec
      Start  4: sign/verify
 4/10 Test  #4: sign/verify ......................   Passed    0.04 sec
      Start  5: TLS
 5/10 Test  #5: TLS ..............................   Passed    0.52 sec
      Start  6: context
 6/10 Test  #6: context ..........................   Passed    0.01 sec
      Start  7: grasshopper
 7/10 Test  #7: grasshopper ......................   Passed    0.04 sec
      Start  8: keyexpimp
 8/10 Test  #8: keyexpimp ........................   Passed    0.01 sec
      Start  9: gost89
 9/10 Test  #9: gost89 ...........................   Passed    0.00 sec
      Start 10: engine
10/10 Test #10: engine ...........................***Failed    0.14 sec

90% tests passed, 1 tests failed out of 10

Total Test time (real) =   0.85 sec

The following tests FAILED:
	 10 - engine (Failed)
Errors while running CTest
make: *** [test] Error 8
```
Here's the log: 
[LastTest.master.log.txt](https://github.com/gost-engine/engine/files/3411624/LastTest.master.log.txt)




Branch `openssl_1_1_0` compiles fine, but fails the same (presumably) test:
```
Running tests...
Test project /Users/ur20980/src/engine/build
    Start 1: engine
1/2 Test #1: engine ...........................***Failed    0.99 sec
    Start 2: grasshopper
2/2 Test #2: grasshopper ......................   Passed    0.01 sec

50% tests passed, 1 tests failed out of 2

Total Test time (real) =   1.01 sec

The following tests FAILED:
	  1 - engine (Failed)
Errors while running CTest
make: *** [test] Error 8
```

Here's the log from `openssl_1_1_0` branch of this repo, compiled with the standard (Macports-installed) OpenSSL_1_1_1c: 
[LastTest.log.txt](https://github.com/gost-engine/engine/files/3411575/LastTest.log.txt)

```
$ env | grep OPENSSL
OPENSSL_LIB_DIR=/opt/local/lib
OPENSSL_INCLUDE_DIR=/opt/local/include
OPENSSL_CFLAGS=-Os -Ofast -std=gnu17 -march=native -I/opt/local/include
OPENSSL_CONF=/opt/local/etc/openssl/openssl.cnf
$ cmake ..
-- The C compiler identification is AppleClang 10.0.1.10010046
-- Check for working C compiler: /opt/local/bin/clang
-- Check for working C compiler: /opt/local/bin/clang -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Using Clang and adding -Qunused-arguments flag
-- Check if the system is big endian
-- Searching 16 bit integer
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of unsigned short
-- Check size of unsigned short - done
-- Using unsigned short
-- Check if the system is big endian - little endian
-- LITTLE_ENDIAN
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/ur20980/src/engine/build
$ make
Scanning dependencies of target gost
[  3%] Building C object CMakeFiles/gost.dir/gost_ec_keyx.c.o
/Users/ur20980/src/engine/gost_ec_keyx.c:10:10: fatal error: 'openssl/evp.h' file not found
#include <openssl/evp.h>
         ^~~~~~~~~~~~~~~
1 error generated.
make[2]: *** [CMakeFiles/gost.dir/gost_ec_keyx.c.o] Error 1
make[1]: *** [CMakeFiles/gost.dir/all] Error 2
make: *** [all] Error 2
$ 
```

Other branches seem to be able to find OpenSSL installation just fine. ???

--- a/gost_pmeth.c
+++ b/gost_pmeth.c
@@ -12,11 +12,18 @@
 #include <openssl/ec.h>
 #include <openssl/err.h>
 #include <openssl/x509v3.h>     /* For string_to_hex */
+#include <openssl/opensslv.h>   /* For OPENSSL_VERSION_MAJOR */
 #include <stdlib.h>
 #include <string.h>
 #include <ctype.h>
 #include "gost_lcl.h"
 #include "e_gost_err.h"
+
+#define ossl3_const
+#ifdef OPENSSL_VERSION_MAJOR
+#undef ossl3_const
+#define ossl3_const const
+#endif
 
 /* -----init, cleanup, copy - uniform for all algs  --------------*/
 /* Allocates new gost_pmeth_data structure and assigns it as data */
@@ -53,7 +60,7 @@ static int pkey_gost_init(EVP_PKEY_CTX *ctx)
 }
 
 /* Copies contents of gost_pmeth_data structure */
-static int pkey_gost_copy(EVP_PKEY_CTX *dst, const EVP_PKEY_CTX *src)
+static int pkey_gost_copy(EVP_PKEY_CTX *dst, ossl3_const EVP_PKEY_CTX *src)
 {
     struct gost_pmeth_data *dst_data, *src_data;
     if (!pkey_gost_init(dst)) {
@@ -564,7 +571,7 @@ static void pkey_gost_mac_cleanup(EVP_PKEY_CTX *ctx)
         OPENSSL_free(data);
 }
 
-static int pkey_gost_mac_copy(EVP_PKEY_CTX *dst, const EVP_PKEY_CTX *src)
+static int pkey_gost_mac_copy(EVP_PKEY_CTX *dst, ossl3_const EVP_PKEY_CTX *src)
 {
     struct gost_mac_pmeth_data *dst_data, *src_data;
     if (!pkey_gost_mac_init(dst)) {

